pkg/virt-launcher/virtwrap/network/ndp/connection.go

```diff
@@ -0,0 +1,126 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package ndp
+
+import (
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/mdlayher/ndp"
+	"golang.org/x/net/ipv6"
+)
+
+const (
+	chkOff  = 2
+	maxHops = 255
+)
+
+// A NDPConnection instruments a system.Conn and adds retry functionality for
+// receiving / sending NDP messages on a given interface.
+type NDPConnection struct {
+	iface      *net.Interface
+	rawConn    *net.IPConn
+	conn       *ipv6.PacketConn
+	controlMsg *ipv6.ControlMessage
+}
+
+// Return an NDPConnection bound to the chosen interface.
+func NewNDPConnection(ifaceName string) (*NDPConnection, error) {
+	iface, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		return nil, fmt.Errorf("could not find interface %s: %v", ifaceName, err)
+	}
+
+	listenAddr := &net.IPAddr{
+		IP:   net.IPv6unspecified,
+		Zone: ifaceName,
+	}
+	icmpListener, err := net.ListenIP("ip6:ipv6-icmp", listenAddr)
```

Comment:
I believe the fd virt-hander opens should also be bound to `ifaceName` (the virt-launcher bridge) - `SO_BINDTODEVICE`.
So that a possible attacker won't be able to somehow abuse the privileged fd virt-launcher has to send fake RAs.

Maybe something similar to [what dhcpv4 does](https://github.com/kubevirt/kubevirt/blob/eafe781f827d6a607edd24efb5d6930410bacdbf/pkg/virt-launcher/virtwrap/network/dhcp/socket_listener.go#L60) can be used to set the `SO_BINDTODEVICE` of the fd.

Reply:
Good news; could not connect to the socket if the interface specified in the `SO_BINDTODEVICE` socket option differs to one you're attempting to connect to.

Used the [following gist](https://gist.github.com/maiqueb/4072d75d75c69513bdd5630a8e96f90b), which has the folloing output:

```bash
I0120 15:20:20.796573       1 main.go:156] Will create a RAW socket on interface: eth0
Error: bind to wrong devico means no conecto !!!: network is unreachable
Usage:
  capabilities-demo bind-experiment [flags]

Flags:
  -h, --help   help for bind-experiment

could *NOT* connect to the socket: network is unreachable

```

Reply:
Thanks for the verification. Does your test pass when you pass "br0" as the argument to the command?
I couldn't find on the gist link where the zone is set on the `saddr`.

Reply:
It doesn't; I was too eager to reply.

Even after using the `SO_BINDTODEVICE` socket option, you can write using that socket to other interfaces.

Moving forward w/ this PR never the less, and will add an option to opt out of the RAs.